src/inmanta/execute/runtime.py

```diff
@@ -337,13 +341,68 @@ def set_value(self, value: ListValue, location: Location, recur: bool = True) ->
             # any set_value may fulfill a promise, allowing this object to be queued
             if self.can_get():
                 self.queue()
-            return
+            return False
 
         self.value.append(value)
 
         for l in self.listeners:
             l.receive_result(value, location)
 
+        return True
+
+    def set_value(self, value: ListValue, location: Location, recur: bool = True) -> None:
+        if not self._set_value(value, location, recur):
+            return
+        if self.can_get():
+            self.queue()
+
+    def can_get(self) -> bool:
+        return self.get_waiting_providers() == 0
+
+    def get_progress_potential(self) -> int:
+        """How many are actually waiting for us """
+        return len(self.waiters) - len(self.listeners)
+
+    def receive_result(self, value: ListValue, location: Location) -> None:
+        self.set_value(value, location)
+
+    def listener(self, resultcollector: ResultCollector, location: Location) -> None:
+        for value in self.value:
+            resultcollector.receive_result(value, location)
+        self.listeners.append(resultcollector)
+
+    def is_multi(self) -> bool:
+        return True
+
+    def __str__(self) -> str:
+        return "BaseListVariable %s" % (self.value)
+
+
+class TempListVariable(BaseListVariable):
+
+    __slots__ = ()
+
+    def set_promised_value(self, promis: Promise, value: ListValue, location: Location, recur: bool = True) -> None:
+        super().set_promised_value(promis, value, location, recur)
+        # 100% accurate promisse tracking
+        if len(self.promisses) == len(self.done_promisses):
```

Comment:
Why is this specific for TempListVariable? It seems like a general rule, applicable to any list. Am I wrong?

Reply:
It is only true if all statements that may add to this list have been emitted, which is, in general, impossible to predict.